### PR TITLE
Flatten the WebRender API to allow us to use shared memory to transfer display lists.

### PR DESCRIPTION
**Pull request by pcwalton**
Improves performance significantly.

r? @glennw

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -24,7 +24,8 @@ pub enum ApiMsg {
     AddNativeFont(FontKey, NativeFontHandle),
     AddImage(ImageKey, u32, u32, ImageFormat, Vec<u8>),
     UpdateImage(ImageKey, u32, u32, ImageFormat, Vec<u8>),
-    AddDisplayList(DisplayListId, PipelineId, Epoch, DisplayListBuilder),
+    AddDisplayList(DisplayListId, PipelineId, Epoch, BuiltDisplayList),
+    AddAuxiliaryLists(PipelineId, AuxiliaryLists),
```
Comment (glennw) [inline src/api.rs]: Are there any potential races here - would it make sense to pass the lists as a field in SetRootStackingContext?

File: src/display_list.rs [hunk 2]
```diff
@@ -342,7 +372,169 @@ impl DisplayListBuilder {
...
+        range.get(&self.display_items)
+    }
+
+    pub fn display_items_mut<'a>(&'a mut self, range: &ItemRange) -> &'a mut [DisplayItem] {
+        range.get_mut(&mut self.display_items)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct ItemRange {
```
Comment (glennw) [inline src/display_list.rs:400]: Doesn't need to block this PR - but perhaps we should newtype these per item type in the future or something similar?

**Comment by pcwalton** [comment]
Updated. r? @glennw

**Pull request merged**